### PR TITLE
New bus (Queue Bus)

### PR DESCRIPTION
**Pull request by evantahler**
Notes: 
- This version does not need to be able to work old driver events (we don't need to be that interoperable)

### CODE REVIEW

**Comment by evantahler** [comment]
Once this is merged in, this project will be moved to https://github.com/queue-bus/node-bus

File: lib/sections/driver.js [hunk 1]
```diff
@@ -1,60 +1,89 @@
...
+        driverPerform(self, args, callback); 
+      }else if(args.bus_class_proxy === '::QueueBus::Rider'){    
+        callback(new Error('TODO: ::QueueBus::Rider')); // TODO!
+      }else if(args.bus_class_proxy === '::QueueBus::Publisher'){
+        publisherPerform(self, args, callback); 
+      }else if(args.bus_class_proxy === '::QueueBus::Heartbeat'){
+        callback(new Error('TODO: ::QueueBus::Heartbeat')); // TODO!
+      }else{
+        callback(new Error('Missing bus_class_proxy or unknown method (' + args.bus_class_proxy + ')'));
+      }
```
Comment (evantahler) [inline lib/sections/driver.js:20]: @bleonard So I know what to do when `args.bus_class_proxy === '::QueueBus::Publisher'` (a normal "fan out" task) and `args.bus_class_proxy === '::QueueBus::Driver'` (a delayed job), but what about:
- `args.bus_class_proxy === '::QueueBus::Rider'`
- `args.bus_class_proxy === '::QueueBus::Heartbeat'`

File: lib/sections/driver.js [hunk 2]
```diff
@@ -1,60 +1,89 @@
...
-  payload['bus_rider_app_key']    = bus_rider_app_key;
-  payload['bus_rider_sub_key']    = bus_rider_sub_key;
-  payload['bus_rider_class_name'] = bus_rider_class_name;
-  payload['bus_event_type']       = bus_event_type;
+  payload.bus_driven_at        = utils.timestamp();
+  payload.bus_rider_queue      = bus_rider_queue;
+  payload.bus_rider_app_key    = bus_rider_app_key;
+  payload.bus_rider_sub_key    = bus_rider_sub_key;
+  payload.bus_rider_class_name = bus_rider_class_name;
+  payload.bus_event_type       = bus_event_type;
```
Comment (evantahler) [inline lib/sections/driver.js:82]: @bleonard from a cursory glance, it seems like the driver metadata has not changed... am I missing anything?

File: lib/sections/publish.js [hunk 3]
```diff
@@ -1,79 +1,62 @@
...
 
 var publish = function(eventType, args, callback){
   var self = this;
   var queue = self.options.incommigQueue;
-  var klass = self.options.busDriverClassKey;
-
-  payload = [publishMetadata(eventType, args)];
+  var klass = self.options.busClassKey;
+  var payload = [publishMetadata(eventType, args)];
+  payload[0].bus_class_proxy  = '::QueueBus::Driver';
```
Comment (evantahler) [inline lib/sections/publish.js]: On a normal publish, `bus_class_proxy` is always `::QueueBus::Driver`

File: lib/sections/publish.js [hunk 4]
```diff
@@ -1,79 +1,62 @@
...
-  var klass = self.options.busPublisherClassKey;
-
-  payload = [publishMetadata(eventType, args)];
-  if(payload["bus_delayed_until"] == null){
-    payload["bus_delayed_until"] = Math.floor(timestamp/1000);
+  var self    = this;
+  var queue   = self.options.incommigQueue;
+  var klass   = self.options.busClassKey;
+  var payload = [publishMetadata(eventType, args)];
+  payload[0].bus_class_proxy  = '::QueueBus::Publisher';
```
Comment (evantahler) [inline lib/sections/publish.js]: On a delay publishAt/publishIn, `bus_class_proxy` is always `::QueueBus:: Publisher`

File: lib/sections/publish.js [hunk 5]
```diff
@@ -1,79 +1,62 @@
...
-  payload['bus_created_at']   = utils.timestamp(); // TODO: get this back in ruby resque-bus
-  payload['bus_published_at'] = utils.timestamp();
-  payload['bus_id']           = utils.timestamp() + "-" + uuid.v4();
-  payload['bus_app_hostname'] = os.hostname();
-  // payload['bus_locale']       = i18n.getLocale();
-  // payload['bus_timezone']     = time.currentTimezone; 
+  payload.bus_created_at   = utils.timestamp(); // TODO: get this back in ruby resque-bus
+  payload.bus_published_at = utils.timestamp();
+  payload.bus_id           = utils.timestamp() + "-" + uuid.v4();
+  payload.bus_app_hostname = os.hostname();
```
Comment (evantahler) [inline lib/sections/publish.js:51]: @bleonard one more check for appended metadata (publish)

File: lib/sections/subscriptions.js [hunk 6]
```diff
@@ -6,50 +6,43 @@ var subscriptions = function(callback){
...
+};
 
 var fullSubscriptionSet = function(self){
   return self.options.connection.namespace + self.options.subscriptionSet;
-}
+};
 
 var rubyizeMatcher = function(matcher){
   for(var i in matcher){
     matcher[i] = utils.toRubyRegExp(matcher[i]);
```
Comment (evantahler) [inline lib/sections/subscriptions.js:42]: BOO!

File: lib/sections/driver.js (hunk 1 above)
Comment (bleonard) [inline lib/sections/driver.js:20]: i think you had them backwards, so to confirm
publisher: publish to incoming later
driver: demultiplex to many other queues
rider: this is what it gets when it is subscribed. you must have been handling this case before.
heartbeat: publish current time data to the incoming queue (this one is not a high priority)

File: lib/sections/driver.js (hunk 2 above)
Comment (bleonard) [inline lib/sections/driver.js:82]: everything is the same except your should insert into the queue using the `bus_class_proxy` pattern.
so you'll want to set `payload.bus_class_proxy = '::QueueBus::Driver'` here and when you push to redis, be sure to push to `bus_incoming` with the worker class being `QueueBus::Worker`
Also pay attention to the args in that redis entry. They are a JSON string now.

File: lib/sections/publish.js (hunk 5 above)
Comment (bleonard) [inline lib/sections/publish.js:51]: here are the fields
https://github.com/queue-bus/queue-bus/blob/f5822ce1da81fd4cae4191c59726cbcd7cbd702d/lib/queue_bus/publishing.rb#L24

bus_published_at
bus_event_type
bus_id
bus_app_hostname
bus_locale
bus_timezone

File: lib/sections/subscriptions.js (hunk 6 above)
Comment (bleonard) [inline lib/sections/subscriptions.js:42]: make an issue (and have a tissue).

File: lib/sections/driver.js (hunk 1 above)
Comment (evantahler) [inline lib/sections/driver.js:20]: Whoops! That was backwards, yes (code was OK).  

I'll keep the heartbeat as a TODO for now, but I'm curious who publishes that first event.  Scheduler?

**Comment by evantahler** [comment]
@bleonard I don't see how the `::QueueBus::Rider` bus_class_proxy is ever needed:
- The driver gets a job with `args.bus_class_proxy = ::QueueBus::Driver` and fans it out.  
  - this moves the job to each new queue, but the new class name [comes from the subscription](https://github.com/taskrabbit/node-resquebus/blob/new_bus/lib/sections/driver.js#L54).  
- Subscriptions look like this: `bus.subscribe(appKey, priority, job, matcher, callback);`
  - I could have a subscription like this: `bus.subscribe('myApp', 'default', 'handleNewUser', { bus_event_type : /^user_created/ });`
  - All new events moved into the new queue would end up looking like: 

``` js
[{"bus_driven_at"=>1426765565, "bus_rider_queue"=>"myApp_default", "bus_rider_app_key"=>"myApp", "bus_rider_sub_key"=>"myApp_default_handleNewUser", "bus_rider_class_name"=>"handleNewUser", "bus_event_type"=>"user_created", "bus_created_at"=>1426765560, "bus_published_at"=>1426765560, "bus_id"=>"1426765560-622a613c-12ae-4fe4-8a9d-14ca8b85525d", "bus_app_hostname"=>"Opsimus-2.local", "user_id"=>5, "bus_class_proxy"=>"::QueueBus::Driver"}]
```

I guess when the driver drives, `"bus_class_proxy"=>"::QueueBus::Driver"` should be changed to "bus_class_proxy"=>"::QueueBus::Rider", (which I changed here: a4d4f7d0c1a7ced692d2c4743a6fe7cd93606b80) but since all the "workers" are normal resque workers, they would never check `bus_class_proxy`.  It would just be present in the `args` hash they get.

I wrote out a whole [example](https://github.com/taskrabbit/node-resquebus/blob/new_bus/example/example.js) you can check.

File: lib/sections/publish.js (hunk 5 above)
Comment (evantahler) [inline lib/sections/publish.js:51]: cool! locale and timezone added here https://github.com/taskrabbit/node-resquebus/commit/9717e0a4ecce3e857d5676a74fb819245019e5fe

File: lib/sections/driver.js (hunk 1 above)
Comment (bleonard) [inline lib/sections/driver.js:20]: It's set up in resque scheduler to run every minute. It also keeps track (in redis) of when it last published by incrementing some number or something and it works on the "now" time. That is, it might publish two or more events if it gets behind.

File: lib/sections/publish.js [hunk 7]
```diff
@@ -1,79 +1,69 @@
...
+  if(eventType){
+    payload.bus_event_type = eventType;
   } else {
-    payload['bus_event_type'] = null;
+    payload.bus_event_type = null;
+  }
+  payload.bus_published_at = utils.timestamp();
+  payload.bus_id           = utils.timestamp() + "-" + uuid.v4();
+  payload.bus_app_hostname = os.hostname();
+  payload.bus_timezone     = new time.Date().getTimezone();
```
Comment (bleonard) [inline lib/sections/publish.js:52]: be interesting to see if this matches up with the ruby strings. if not, we might want to leave it off.

**Comment by bleonard** [comment]
The driver looks at the subscriptions and enqueues and new item in resque/sidekiq with the class of `::QueueBus::Worker`. That worker has a class_proxy of `::QueueBus::Rider` and args about the subscription id and method and such (`bus_rider_sub_key` and `bus_rider_app_key`).

The code you pointed to (https://github.com/taskrabbit/node-resquebus/blob/new_bus/lib/sections/driver.js#L54) doesn't look like it's queueing up a `Worker` with a proxy. `subscription.class` should usually be `QueueBus::Rider` so I would expect that code to look more like:

``` javascript
var payload = driverMetadata(args, subscription.queue_name, app, i, subscription.class, args.bus_event_type);
payload.bus_class_proxy = subscription.class;
self.queueObject.enqueue(subscription.queue_name, '::QueueBus::Worker', payload, function(err, toRun){
```

The args you have there are not right if those are the ones that got fanned out because you have `"bus_class_proxy"=>"::QueueBus::Driver"` - that would likely be`Rider` if the subscription was from the normal subscribing mechanism. Ok, you noted that it seems. What you are missing is that you should enqueue it with the class of `QueueBus::Worker`

To recap:
So Resque pops the item from that queue sometime later and sees that it's a `::Worker` so it runs the perform method on that. `Worker` knows to proxy to the class it is given. In this case it is usually `Rider`. `Rider` knows how to use the `bus_rider_sub_key` and `bus_rider_app_key` to execute the right code.

File: lib/sections/driver.js [hunk 8]
```diff
@@ -1,60 +1,88 @@
...
-                  if(started === 0 && typeof callback === 'function'){ callback(null, matches); }
-                });
-              }else{
-                process.nextTick(function(){
-                  started--;
-                  if(started === 0 && typeof callback === 'function'){ callback(null, matches); }
-                });
-              }
-            }
+      if(args.bus_class_proxy === '::QueueBus::Driver'){
```
Comment (bleonard) [inline lib/sections/driver.js]: for safety, it would be best to also handle these without the colons. e.g. `QueueBus::Driver`

File: lib/sections/driver.js [hunk 9]
```diff
@@ -1,60 +1,88 @@
...
-                process.nextTick(function(){
-                  started--;
-                  if(started === 0 && typeof callback === 'function'){ callback(null, matches); }
-                });
-              }
-            }
+      if(args.bus_class_proxy === '::QueueBus::Driver'){   
+        driverPerform(self, args, callback); 
+      }else if(args.bus_class_proxy === '::QueueBus::Rider'){    
+        callback(new Error('TODO: ::QueueBus::Rider')); // TODO!
```
Comment (bleonard) [inline lib/sections/driver.js]: Here's the code for rider.
https://github.com/queue-bus/queue-bus/blob/5c009a3b2d2d58994813bdca0dc78547a18b0295/lib/queue_bus/rider.rb

it basically looks up the function to execute from `bus_rider_sub_key` and `bus_rider_app_key`
You must have had this working in the current version, otherwise nothing would work. That is, it would never have received anything that was driven to it.
